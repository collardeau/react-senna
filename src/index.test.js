import React from "react";
import Senna from "./index";
import renderer from "react-test-renderer";

function renderComp(props = {}) {
  return renderer.create(<Senna {...props} />);
}

function getProps(props) {
  const comp = renderComp(props);
  const tree = comp.toJSON();
  return tree && tree.props;
}

describe("Props Handling", () => {
  test("no props given", () => {
    expect(getProps()).toMatchSnapshot();
  });

  test("passes user props", () => {
    expect(getProps({ a: "thing" }).a).toBe("thing");
  });

  test("basic case", () => {
    expect(
      getProps({
        seeds: [{ name: "users" }]
      })
    ).toMatchSnapshot();
  });

  test("normal case", () => {
    expect(
      getProps({
        seeds: [
          { name: "users", initialState: {}, loadable: true },
          { name: "isVisible", initialState: false, toggleable: true }
        ]
      })
    ).toMatchSnapshot();
  });
});

describe("Flags", () => {
  test("loaded flag", () => {
    const comp = renderComp({
      seeds: [
        {
          name: "users",
          initialState: {},
          loadable: true
        }
      ]
    });
    let tree = comp.toJSON();
    expect(tree.props).toMatchSnapshot();
    expect(tree.props.usersLoaded).toBe(false);
    tree.props.handlers.setUsers({ uid: "some-uid" });
    tree = comp.toJSON();
    expect(tree.props.usersLoaded).toBe(true);
  });

  test("mergeable with objects", () => {
    const comp = renderComp({
      seeds: [
        {
          name: "users",
          initialState: {},
          mergeable: true
        }
      ]
    });
    let tree = comp.toJSON();
    tree.props.handlers.setUsers({ a: "a" });
    tree.props.handlers.mergeUsers({ b: "b" });
    tree = comp.toJSON();
    expect(tree.props.users.a).toBe("a");
    expect(tree.props.users.b).toBe("b");
  });

  test("mergeable with arrays", () => {
    const comp = renderComp({
      seeds: [
        {
          name: "users",
          initialState: [],
          mergeable: true
        }
      ]
    });
    let tree = comp.toJSON();
    tree.props.handlers.setUsers(["a"]);
    tree.props.handlers.mergeUsers(["b"]);
    tree = comp.toJSON();
    expect(tree.props.users[0]).toBe("a");
    expect(tree.props.users[1]).toBe("b");
  });
});
describe("Errors", () => {
  test("mergeable with num for initial state", () => {
    const spy = jest.fn();
    const comp = renderComp({
      seeds: [
        {
          name: "x",
          initialState: 0,
          mergeable: true
        }
      ],
      onError: spy
    });
    expect(spy).toBeCalled();
  });
  test("mergeable with string for initial state", () => {
    const spy = jest.fn();
    const comp = renderComp({
      seeds: [
        {
          name: "users",
          initialState: "hi",
          mergeable: true
        }
      ],
      onError: spy
    });
    expect(spy).toBeCalled();
  });
  test("set a mergeable state to a num", () => {
    const spy = jest.fn();
    const comp = renderComp({
      seeds: [
        {
          name: "users",
          initialState: {},
          mergeable: true
        }
      ],
      onError: spy
    });
    const tree = comp.toJSON();
    tree.props.handlers.setUsers(0);
    expect(spy).toBeCalled();
  });
  test("set a mergeable state to a string", () => {
    const spy = jest.fn();
    const comp = renderComp({
      seeds: [
        {
          name: "users",
          initialState: {},
          mergeable: true
        }
      ],
      onError: spy
    });
    const tree = comp.toJSON();
    tree.props.handlers.setUsers("hi");
    expect(spy).toBeCalled();
  });
});

describe("Use cases", () => {
  test("toggle example", () => {
    const comp = renderComp({
      seeds: [
        {
          name: "active",
          initialState: false,
          toggleable: true
        }
      ]
    });
    let tree = comp.toJSON();
    expect(tree.props.active).toBe(false);
    tree.props.handlers.toggleActive();
    tree = comp.toJSON();
    expect(tree.props.active).toBe(true);
    expect(tree.props).toMatchSnapshot();
    tree.props.handlers.toggleActive();
    tree = comp.toJSON();
    expect(tree.props.active).toBe(false);
  });

  test("counter example", () => {
    const comp = renderComp({
      seeds: [
        {
          name: "count",
          initialState: 0,
          resetable: true,
          handlers: {
            incr: st => st + 1
          }
        }
      ]
    });
    let tree = comp.toJSON();
    expect(tree.props.count).toBe(0);
    tree.props.handlers.incrCount();
    tree = comp.toJSON();
    expect(tree.props.count).toBe(1);
    expect(tree.props).toMatchSnapshot();
    tree.props.handlers.resetCount();
    tree = comp.toJSON();
    expect(tree.props.count).toBe(0);
  });

  test("with compound handlers example", () => {
    const comp = renderComp({
      seeds: [
        {
          name: "countA",
          initialState: 0
        },
        {
          name: "countB",
          initialState: 0
        }
      ],
      withHandlers: {
        setAll: props => num => {
          props.handlers.setCountA(num);
          props.handlers.setCountB(num);
        }
      }
    });
    let tree = comp.toJSON();
    expect(tree.props).toMatchSnapshot();
    tree.props.handlers.setAll(10);
    tree = comp.toJSON();
    expect(tree.props.countA).toBe(10);
    expect(tree.props.countB).toBe(10);
  });
});
