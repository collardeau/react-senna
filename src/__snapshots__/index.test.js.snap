// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic case 1`] = `
Object {
  "handlers": Object {
    "setUsers": [Function],
  },
  "users": undefined,
}
`;

exports[`counter example 1`] = `
Object {
  "count": 1,
  "handlers": Object {
    "incrCount": [Function],
    "resetCount": [Function],
    "setCount": [Function],
  },
}
`;

exports[`loaded flag 1`] = `
Object {
  "handlers": Object {
    "setUsers": [Function],
    "setUsersLoaded": [Function],
  },
  "users": Object {},
  "usersLoaded": false,
}
`;

exports[`normal case 1`] = `
Object {
  "handlers": Object {
    "setIsVisible": [Function],
    "setUsers": [Function],
    "setUsersLoaded": [Function],
    "toggleIsVisible": [Function],
  },
  "isVisible": false,
  "users": Object {},
  "usersLoaded": false,
}
`;

exports[`toggle example 1`] = `
Object {
  "active": true,
  "handlers": Object {
    "setActive": [Function],
    "toggleActive": [Function],
  },
}
`;

exports[`withHandlers example 1`] = `
Object {
  "countA": 0,
  "countB": 0,
  "handlers": Object {
    "setAll": [Function],
    "setCountA": [Function],
    "setCountB": [Function],
  },
}
`;

exports[`without any config 1`] = `
Object {
  "handlers": Object {},
}
`;
